cartographer_grpc/framework/rpc.h

```diff
@@ -121,11 +128,12 @@ class ActiveRpcs {
   ~ActiveRpcs() EXCLUDES(lock_);
 
   Rpc* Add(std::unique_ptr<Rpc> rpc) EXCLUDES(lock_);
+  std::shared_ptr<Rpc> Get(Rpc* rpc) EXCLUDES(lock_);
   bool Remove(Rpc* rpc) EXCLUDES(lock_);
 
  private:
   cartographer::common::Mutex lock_;
-  std::unordered_set<Rpc*> rpcs_;
+  std::unordered_map<Rpc*, std::shared_ptr<Rpc>> rpcs_;
```

Comment:
This complex data structure shouldn't be necessary for this behavior.
Rpc should create shared_ptr itself.

Reply:
I am not sure what you mean by "Rpc should create shared_ptr itself". `shared_ptr`s form cliques. So you can't create `shared_ptr`s for an object by a member function like this 

~~~
shared_ptr<Rpc> Rpc::GetSharedPtr() {
  return shared_ptr<Rpc>(this);
}
~~~
Creating a shared pointer by calling `shared_ptr<Rpc>(rpc)` starts a new clique and moves ownership of `rpc` to the clique. You extend the clique by making copies of the shared pointer. If you created another shared pointer with `shared_ptr<Rpc>(rpc)` you would get a double free when the second shared_ptr goes out of scope.

Reply:
Ok, I see that my suggestion was not a good idea. There could be an alternative around std::enable_shared_from_this, but that also sounds fishy. Let's stick to the current solution.